plugins/amqp_trx_plugin/include/eosio/amqp_trx_plugin/fifo_trx_processing_queue.hpp

```diff
@@ -0,0 +1,212 @@
+};
+
+} // detail
+
+/**
+ * FIFO queue for starting signature recovery. Enforces limit (max_depth) on queued transactions to be processed.
+ * Only one transaction at a time is submitted to the main application thread so that order is preserved.
+ */
+template<typename ProducerPlugin>
+class fifo_trx_processing_queue : public std::enable_shared_from_this<fifo_trx_processing_queue<ProducerPlugin>> {
+
+private:
+   struct q_item {
+      chain::recover_keys_future fut;
+      producer_plugin::next_function<chain::transaction_trace_ptr> next;
+   };
+   detail::blocking_queue<q_item> queue_;
+   std::thread thread_;
+   std::atomic_bool running_ = true;
+   chain::chain_id_type chain_id_;
+   uint32_t configured_subjective_signature_length_limit_ = 0;
+   boost::asio::io_context& sig_thread_pool_;
+   // Doesn't look like we have strong guarantee that producer_plugin will live longer than app().post() call
+   // but various other places also use producer_plugin* & chain_plugin* in app().post().
+   ProducerPlugin* prod_plugin_ = nullptr;
+
+public:
+
+   /**
+    * @param chain_id for signature recovery
+    * @param configured_subjective_signature_length_limit for signature recovery
+    * @param sig_thread_pool for signature recovery
+    * @param prod_plugin producer_plugin
+    * @param max_depth the queue depth for number of trxs to start signature recovery on
+    */
+   fifo_trx_processing_queue( const chain::chain_id_type& chain_id,
+                              uint32_t configured_subjective_signature_length_limit,
+                              boost::asio::io_context& sig_thread_pool,
+                              ProducerPlugin* prod_plugin,
+                              size_t max_depth )
+   : queue_(max_depth)
+   , chain_id_(chain_id)
+   , configured_subjective_signature_length_limit_(configured_subjective_signature_length_limit)
+   , sig_thread_pool_(sig_thread_pool)
+   , prod_plugin_(prod_plugin)
+   {
+   }
+
+   /// separate run() because of shared_from_this
+   void run() {
+      thread_ = std::thread([self=this->shared_from_this()]() {
+         fc::set_os_thread_name( "trxq" );
+         while( self->running_ ) {
+            try {
+               q_item i;
+               if( self->queue_.pop_and_pause(i) ) {
+                  auto trx_meta = i.fut.get();
+                  dlog("posting trx: ${id}", ("id", trx_meta->id()));
+                  app().post(priority::low, [self, trx{std::move(trx_meta)}, next{std::move(i.next)}](){
+                     auto retry_later = [self](const chain::transaction_metadata_ptr& trx, producer_plugin::next_function<chain::transaction_trace_ptr>& next) {
+                        std::promise<chain::transaction_metadata_ptr> p;
+                        q_item i;
+                        i.fut = p.get_future();
+                        p.set_value(trx);
+                        i.next = next;
+                        self->queue_.push_front( std::move( i ) );
```

Comment:
Is this busy wait okay?

Reply:
If configured for 500ms `max_block_cpu_usage`, then yes. However, I need to put some thought to what to do if configured differently.

Reply:
It now respects cpu effort percent and does not have a busy wait.